### PR TITLE
change `hir::ExprLit` to take a `ConstVal` instead of an `ast::Lit`

### PR DESCRIPTION
**Pull request by oli-obk**
This PR uses the lowering step to move from `ast::Lit` to `ConstVal`. Since suffixed literals (e.g. `42_u8`) are turned into concrete `ConstInt` variants, these literal overflow checks need to be done in the ast. This causes a little code duplication in the `rustc_lint/types` lints.

~~I don't know how immutable the HIR is, but if we can run a mutating pass after typeck, we could also lower all `ConstInt::Infer`/`ConstInt::InferSigned` variants to their concrete variants.~~

r? @eddyb 

As a next step I can remove some code-duplication between MIR-trans and normal trans (since MIR already uses `ConstVal` instead of literals).

### CODE REVIEW

File: src/librustc/hir/lowering.rs [hunk 1]
```diff
@@ -1674,6 +1688,121 @@ pub fn lower_expr(lctx: &LoweringContext, e: &Expr) -> P<hir::Expr> {
...
+        (n, Signed(Is), 64) => Integral(ConstInt::Isize(ConstIsize::Is64(-(n as i64)))),
+        (n, Signed(Is), 32) => Integral(ConstInt::Isize(ConstIsize::Is32(-(n as i64 as i32)))),
+        (_, Signed(Is), _) => unreachable!(),
+         // unary negation of unsigned has already been reported by EarlyTypeLimits
+        (_, Unsigned(_), _) => ConstVal::Dummy,
+        (n, Unsuffixed, _) => Integral(ConstInt::InferSigned(-(n as i64))),
+    }
+}
+
+pub fn lower_un_neg(lctx: &LoweringContext, inner: &Expr) -> (hir::Expr_, ThinAttributes) {
```
Comment (eddyb) [inline src/librustc/hir/lowering.rs]: Is all of this necessary? Can't we just keep the negation expression?

**Comment by arielb1** [comment]
I am quite sure the HIR is supposed to be immutable

**Comment by eddyb** [comment]
I think everything which wants accurate information should used MIR as the source of truth - maybe the lints could be triggered from the MIR building code?

File: src/librustc/hir/lowering.rs (hunk 1 above)
Comment (arielb1) [inline src/librustc/hir/lowering.rs]: This works better in patterns

**Comment by oli-obk** [comment]
@eddyb triggering lints manually is weird

I have a different suggestion:

* during AST -> HIR lowering do the following conversions

| AST | HIR | breaking change? |
|-----|-----|----|
| LitKind::Int(n, Unsuffixed) | ConstInt::Infer(n) | - |
| ExprKind::UnOp(Neg, LitKind::Int(n, Unsuffixed)) | ConstInt::InferSigned(-n) | Yes, for negated literals bigger than `i64::MAX`. Note, we already error for all literals bigger than `u64::MAX` |
| LitKind::Int(n, t) | ExprType(ConstInt::Infer(n), t) | - |
| ExprKind::UnOp(Neg, LitKind::Int(n, t)) | ExprType(ConstInt::InferSigned(-n), t)) | Yes, see above |

* during HIR -> MIR lowering do nothing
* add a MIR-pass that folds numeric constants and lints on bad things.

**Comment by eddyb** [comment]
@oli-obk You can't really lint after you've built the MIR in the same way because you lose location information (spans are there, but no IDs to get lint attributes for - maybe we need to track lints some other way? in spans? cc @nrc).

File: src/librustc_const_eval/eval.rs [hunk 2]
```diff
@@ -1145,10 +1152,17 @@ fn lit_to_const<'tcx>(lit: &ast::LitKind,
             infer(Infer(n), tcx, &ty::TyUint(ity), span).map(Integral)
         },
 
-        LitKind::Float(ref n, _) |
+        LitKind::Float(ref n, fty) => {
+            if let Ok(x) = n.parse::<f64>() {
```
Comment (arielb1) [inline src/librustc_const_eval/eval.rs]: Is this *supposed* to do a broken double-rounding?

Comment (rkruppe) [inline src/librustc_const_eval/eval.rs]: I don't know why anyone would *want* to do that, but `ConstVal` doesn't have separate variants for f64 and f32 so avoiding the double rounding doesn't seem possible without semi-large changes all throughout const eval. I've been meaning to find a short example program that demonstrates the double-rounding and file an issue. In any case, fixing this existing brokenness seems orthogonal to and out of scope for this PR.

File: src/librustc/hir/lowering.rs [hunk 3]
```diff
@@ -1674,6 +1676,63 @@ pub fn lower_expr(lctx: &LoweringContext, e: &Expr) -> P<hir::Expr> {
...
+        Int(n, Unsigned(Us)) => {
+            ConstVal::Integral(match lctx.id_assigner.target_bitwidth() {
+                32 => ConstInt::Usize(ConstUsize::Us32(n as u32)),
+                64 => ConstInt::Usize(ConstUsize::Us64(n)),
+                _ => unimplemented!(),
+            })
+        },
+        Int(n, Unsuffixed) => ConstVal::Integral(ConstInt::Infer(n)),
+        Float(ref f, fty) => {
+            if let Ok(x) = f.parse::<f64>() {
```
Comment (arielb1) [inline src/librustc/hir/lowering.rs]: double rounding

File: src/librustc_const_eval/eval.rs (hunk 2 above)
Comment (arielb1) [inline src/librustc_const_eval/eval.rs]: There's #32805 which is vaguely relevant.

**Comment by oli-obk** [comment]
@eddyb well I could do [these changes](https://github.com/rust-lang/rust/pull/32793#issuecomment-206920966) without touching MIR. I'll simply reinstate the current HIR-lint.

**Comment by oli-obk** [comment]
closing until things have settled around lints and MIR

**Pull request closed without merging**